### ISSUE TITLE
clipboard workaround

### ISSUE THREAD

**Issue by ngmariusz** [jeffreytse/zsh-vi-mode#19]
While we wait for final solution you can patch zsh-vi-mode plugin if it's deal breaker for you.

For those not very familiar with bash just find plugin file `~/.oh-my-zsh/custom/plugins/zsh-vi-mode/zsh-vi-mode.zsh` and then function `zvm_yank()`; In my case for windows wsl2 ubuntu20 it's like so:

![image](https://user-images.githubusercontent.com/3337832/108175109-a0fff600-7100-11eb-9258-1521791a4b69.png)

brutal, but does the job.

**Comment by tormodatt** [comment]
similarly for macOS:
```zsh
function zvm_vi_yank() {
	zvm_yank
	echo ${CUTBUFFER} | pbcopy
	zvm_exit_visual_mode
}
```
(printf also works)

**Comment by jeffreytse** [comment]
@tormodatt Thanks for your tips. Hope you have a good day! : )

**Comment by vanarok** [comment]
Also for wayland (need is wl-clipboard)
zvm_vi_yank () {
	zvm_yank
	printf %s "${CUTBUFFER}" |  wl-copy -n
	zvm_exit_visual_mode
}

**Comment by reaz1995** [comment]
also you can make p/P  paste from the windows clipboard by changing zvm_vi_put_after and zvm_vi_put_before as follow:
```zsh
function zvm_vi_put_after() {
  local head= foot=
  local content=$(win32yank.exe -o)
...
}
```
would be great to add one variable which will decide if use custom clipboards, if yes then left the user specify yank/past command

**Comment by anderslundstedt** [comment]
FWIW, here is my solution for getting zsh-vi-mode to play nicely—given my preferences¹—with system clipboards. I have not done any thorough testing but I have not noticed any problems since I started using this solution some weeks ago. (`cbread` and `cbprint` are custom functions I have for writing and reading to and from the system clipboard, respectively.²)

```
my_zvm_vi_yank() {
  zvm_vi_yank
  echo -en "${CUTBUFFER}" | cbread
}

my_zvm_vi_delete() {
  zvm_vi_delete
  echo -en "${CUTBUFFER}" | cbread
}

my_zvm_vi_change() {
  zvm_vi_change
  echo -en "${CUTBUFFER}" | cbread
}

my_zvm_vi_change_eol() {
  zvm_vi_change_eol
  echo -en "${CUTBUFFER}" | cbread
}

my_zvm_vi_put_after() {
  CUTBUFFER=$(cbprint)
  zvm_vi_put_after
  zvm_highlight clear # zvm_vi_put_after introduces weird highlighting for me
}

my_zvm_vi_put_before() {
  CUTBUFFER=$(cbprint)
  zvm_vi_put_before
  zvm_highlight clear # zvm_vi_put_before introduces weird highlighting for me
}

zvm_after_lazy_keybindings() {
  zvm_define_widget my_zvm_vi_yank
  zvm_define_widget my_zvm_vi_delete
  zvm_define_widget my_zvm_vi_change
  zvm_define_widget my_zvm_vi_change_eol
  zvm_define_widget my_zvm_vi_put_after
  zvm_define_widget my_zvm_vi_put_before

  zvm_bindkey visual 'y' my_zvm_vi_yank
  zvm_bindkey visual 'd' my_zvm_vi_delete
  zvm_bindkey visual 'x' my_zvm_vi_delete
  zvm_bindkey vicmd  'C' my_zvm_vi_change_eol
  zvm_bindkey visual 'c' my_zvm_vi_change
  zvm_bindkey vicmd  'p' my_zvm_vi_put_after
  zvm_bindkey vicmd  'P' my_zvm_vi_put_before
}
```

¹ I only work with one clipboard, which is shared between my local system and any (trusted) remote system I am logged on to.

² Here are `cbread` and `cbprint`:

```
if [[ $(uname) = "Darwin" ]]; then
  on_mac_os=0
else
  on_mac_os=1
fi

cbread() {
  if [[ $on_mac_os -eq 0 ]]; then
    pbcopy
  else
    xclip -selection primary -i -f | xclip -selection secondary -i -f | xclip -selection clipboard -i
  fi
}

cbprint() {
  if [[ $on_mac_os -eq 0 ]]; then
    pbpaste
  else
    if   x=$(xclip -o -selection clipboard 2> /dev/null); then
      echo -n $x
    elif x=$(xclip -o -selection primary   2> /dev/null); then
      echo -n $x
    elif x=$(xclip -o -selection secondary 2> /dev/null); then
      echo -n $x
    fi
  fi
}
```

**Comment by Acumane** [comment]
For anyone on WSL2 for Windows:
```zsh
zvm_vi_yank () {
	zvm_yank
	echo "$CUTBUFFER" | clip.exe
	zvm_exit_visual_mode
}
```

**Comment by raph90** [comment]
Thanks for your solution @anderslundstedt

**Comment by phortonssf** [comment]
Any idea how to stop deleting lines or characters from replacing what is in the clipboard?

**Comment by Shuraken007** [comment]
By some reason zvm_bindkey stopped work. I rewrite this with redefining functions.
```bash
for f in zvm_backward_kill_region zvm_yank zvm_replace_selection zvm_change_surround_text_object zvm_vi_delete zvm_vi_change zvm_vi_change_eol; do
  eval "$(echo "_$f() {"; declare -f $f | tail -n +2)"
  eval "$f() { _$f; echo -en \$CUTBUFFER | cbread }"
done

for f in zvm_vi_put_after zvm_vi_put_before; do
  eval "$(echo "_$f() {"; declare -f $f | tail -n +2)"
  eval "$f() { CUTBUFFER=\$(cbprint); _$f; zvm_highlight clear }"
done
```

**Comment by uduse** [comment]
Sharing my version for macOs that covers more commands.

```
my_zvm_vi_yank() {
    zvm_vi_yank
    echo -en "${CUTBUFFER}" | pbcopy
}

my_zvm_vi_delete() {
    zvm_vi_delete
    echo -en "${CUTBUFFER}" | pbcopy
}

my_zvm_vi_change() {
    zvm_vi_change
    echo -en "${CUTBUFFER}" | pbcopy
}

my_zvm_vi_change_eol() {
    zvm_vi_change_eol
    echo -en "${CUTBUFFER}" | pbcopy
}

my_zvm_vi_substitute() {
    zvm_vi_substitute
    echo -en "${CUTBUFFER}" | pbcopy
}

my_zvm_vi_substitute_whole_line() {
    zvm_vi_substitute_whole_line
    echo -en "${CUTBUFFER}" | pbcopy
}

my_zvm_vi_put_after() {
    CUTBUFFER=$(pbpaste)
    zvm_vi_put_after
    zvm_highlight clear # zvm_vi_put_after introduces weird highlighting
}

my_zvm_vi_put_before() {
    CUTBUFFER=$(pbpaste)
    zvm_vi_put_before
    zvm_highlight clear # zvm_vi_put_before introduces weird highlighting
}

my_zvm_vi_replace_selection() {
    CUTBUFFER=$(pbpaste)
    zvm_vi_replace_selection
    echo -en "${CUTBUFFER}" | pbcopy
}

zvm_after_lazy_keybindings() {
    zvm_define_widget my_zvm_vi_yank
    zvm_define_widget my_zvm_vi_delete
    zvm_define_widget my_zvm_vi_change
    zvm_define_widget my_zvm_vi_change_eol
    zvm_define_widget my_zvm_vi_put_after
    zvm_define_widget my_zvm_vi_put_before
    zvm_define_widget my_zvm_vi_substitute
    zvm_define_widget my_zvm_vi_substitute_whole_line
    zvm_define_widget my_zvm_vi_replace_selection

    zvm_bindkey vicmd 'C' my_zvm_vi_change_eol
    zvm_bindkey vicmd 'P' my_zvm_vi_put_before
    zvm_bindkey vicmd 'S' my_zvm_vi_substitute_whole_line
    zvm_bindkey vicmd 'p' my_zvm_vi_put_after

    zvm_bindkey visual 'p' my_zvm_vi_replace_selection
    zvm_bindkey visual 'c' my_zvm_vi_change
    zvm_bindkey visual 'd' my_zvm_vi_delete
    zvm_bindkey visual 's' my_zvm_vi_substitute
    zvm_bindkey visual 'x' my_zvm_vi_delete
    zvm_bindkey visual 'y' my_zvm_vi_yank
}
```

**Comment by d9k** [comment]
https://github.com/jeffreytse/zsh-vi-mode/issues/19#issuecomment-1268057812:

in Linux for this bindings to work set aliases before:

```bash
alias cbread='xclip -selection c'
alias cbprint='xclip -o -selection clipboard'
```

**Comment by joihn** [comment]
I applied it on linux.
it works well when I first select the word via visual mode, then `y`
however 
- `yy` or `Y` (copy the whole line) does not work
-`yiw` (copy current word) does not work
any idea ?

**Comment by d9k** [comment]
@joihn, thanks!

[permanent link](https://github.com/jeffreytse/zsh-vi-mode/blob/cd730cd347dcc0d8ce1697f67714a90f07da26ed/zsh-vi-mode.zsh#L1829):

> `y*) zvm_vi_yank false; cursor=;;`

**Comment by maxmax1992** [comment]
FYI if you're defining that in your ~/.zshrc and osing oh-my-zsh do it after this line!
`source $ZSH/oh-my-zsh.sh`
Otherwise it wouldn't work!

**Comment by VytautasT** [comment]
A slight improvement on  @Shuraken007 [solution](https://github.com/jeffreytse/zsh-vi-mode/issues/19#issuecomment-1597553775) so that the functions which expect arguments receive them. Was needed for `zvm_vi_replace_selection` to work.

```zsh
for f in zvm_backward_kill_region zvm_yank zvm_replace_selection zvm_change_surround_text_object zvm_vi_delete zvm_vi_change zvm_vi_change_eol; do
  eval "$(echo "_$f() {"; declare -f $f | tail -n +2)"
  eval "$f() { _$f \"\$@\"; echo -en \$CUTBUFFER | cbread }"
done

for f in zvm_vi_put_after zvm_vi_put_before zvm_vi_replace_selection; do
  eval "$(echo "_$f() {"; declare -f $f | tail -n +2)"
  eval "$f() { CUTBUFFER=\$(cbprint); _$f \"\$@\"; zvm_highlight clear }"
done
```

**Comment by semi710** [comment]
host is macos, and remote is linux server.

using zsh-vi-mode on linux, yanking does not send the content to mac's clipboard with tmux i am using osc52 that is working fine, is there any way to get it working

i have also tried setting the below script as copy cmd

```bash
    printf "\033]52;c;%s\007" "$(base64 | tr -d '\n')"
```

```sh
echo "something" | copy.script
```

copies content from remote to host clipboard
but 
```
󰘧 echo $ZVM_CLIPBOARD_COPY_CMD
/nix/store/720wsih34mihbf4la0wz4ddkxjnq4m5m-copy/bin/copy
```
this fails to work

**Comment by jeffreytse** [comment]
Hi @niksingh710 , could u try below solution instead?

```sh
# Install osc command first
go install -v github.com/theimpostor/osc@latest
# Try whether it works
echo -n "asdf" | osc copy
# Try to config with osc command
ZVM_CLIPBOARD_COPY_CMD="osc copy"
```
If you want to only depend on pure shell script, what you need to do is to learn about tmux escape sequence.
FYI. Hope it helpful -> https://gist.github.com/carloseberhardt/d9433f308e31c3ffebc34d07da074e22

**Comment by semi710** [comment]
Hey @jeffreytse,

* Running:

  ```
  echo -n "asdf" | osc copy
  ```

  * Works **outside tmux** (string is copied to clipboard).
  * **Does not work inside tmux** (nothing ends up in clipboard).

* I tried setting:

  ```
  export ZVM_CLIPBOARD_COPY_CMD="osc copy"
  ```

  both inside tmux and outside tmux. ZVM copy still does **not** put anything in the clipboard.

* I also tested the standalone OSC 52 script you attached, and **that script works inside tmux**, so OSC 52 passthrough in tmux is working on my setup.

* I also tried setting `ZVM_CLIPBOARD_COPY_CMD` to that script, but ZVM copy still fails.

So in summary:

* OSC 52 works in tmux (verified with the standalone script).
* `osc copy` works outside tmux.
* `osc copy` does **not** work inside tmux.
* ZVM copy using `ZVM_CLIPBOARD_COPY_CMD` also does **not** work.

**Comment by jeffreytse** [comment]
@niksingh710 Thanks for your information. This could be a potential issue, I will work on a new release for this, welcome to star this project for the new release. : )
